### ISSUE TITLE
testsuite failure on i386 "QuickInspectorTest::testFetchingPreview() Compared values are not the same"

### ISSUE THREAD

**Issue by LocutusOfBorg** [KDAB/GammaRay#226]
This is happening on Debian and Ubuntu i386 machines.

I guess this is something similar to Windows, because I see the test is disabled there:
```
         QCOMPARE(img.width(), 320);
         QCOMPARE(img.height(), 160);
 #ifndef Q_OS_WIN // this is too unstable on the CI, rendered results seem to differ in color!?
         QCOMPARE(img.pixel(1,1), QColor(QStringLiteral("lightsteelblue")).rgb());
 #endif
     }
```
removing that line makes the build fail in some other places
```
xvfb-run -a -s "-screen 0 640x480x24" dh_auto_test --max-parallel=1 -B obj-qt5
	make -j1 test ARGS\+=-j1
make[2]: Entering directory '/build/gammaray-2.4.1/obj-qt5'
Running tests...
/usr/bin/ctest --force-new-ctest-process -j1
Test project /build/gammaray-2.4.1/obj-qt5
      Start  1: connectiontest-preload
 1/21 Test  #1: connectiontest-preload ...........***Failed    0.03 sec
No probe found for ABI qt5.5-i686

      Start  2: connectiontest-style
 2/21 Test  #2: connectiontest-style .............***Failed    0.04 sec
No probe found for ABI qt5.5-i686

      Start  3: connectiontest-gdb
 3/21 Test  #3: connectiontest-gdb ...............***Failed    0.04 sec
No probe found for ABI qt5.5-i686

      Start  4: attachtest-gdb
 4/21 Test  #4: attachtest-gdb ...................***Exception: Other  0.39 sec
attaching gammaray in 380 ms
********* Start testing of TestMain *********
Config: Using QtTest library 5.5.1, Qt 5.5.1 (i386-little_endian-ilp32 shared (dynamic) release build; by GCC 5.3.1 20160509)
PASS   : TestMain::initTestCase()
attaching gammaray
No probe found for ABI qt5.5-i686
could not attach to debuggee

      Start  5: multisignalmappertest
 5/21 Test  #5: multisignalmappertest ............   Passed    0.03 sec
      Start  6: probeabitest
 6/21 Test  #6: probeabitest .....................   Passed    0.03 sec
      Start  7: probeabidetectortest
 7/21 Test  #7: probeabidetectortest .............***Failed    0.04 sec
********* Start testing of ProbeABIDetectorTest *********
Config: Using QtTest library 5.5.1, Qt 5.5.1 (i386-little_endian-ilp32 shared (dynamic) release build; by GCC 5.3.1 20160509)
PASS   : ProbeABIDetectorTest::initTestCase()
FAIL!  : ProbeABIDetectorTest::testDetectExecutable() Compared values are not the same
   Actual   (abi.id())                          : "qt5.5-i686"
   Expected (QStringLiteral(GAMMARAY_PROBE_ABI)): "qt5.5-x86_64"
   Loc: [/build/gammaray-2.4.1/tests/probeabidetectortest.cpp(48)]
FAIL!  : ProbeABIDetectorTest::testDetectProcess() Compared values are not the same
   Actual   (abi.id())                          : "qt5.5-i686"
   Expected (QStringLiteral(GAMMARAY_PROBE_ABI)): "qt5.5-x86_64"
   Loc: [/build/gammaray-2.4.1/tests/probeabidetectortest.cpp(56)]
PASS   : ProbeABIDetectorTest::testContainsQtCore(empty)
PASS   : ProbeABIDetectorTest::testContainsQtCore(too short1)
PASS   : ProbeABIDetectorTest::testContainsQtCore(too short2)
PASS   : ProbeABIDetectorTest::testContainsQtCore(unix1)
PASS   : ProbeABIDetectorTest::testContainsQtCore(unix2)
PASS   : ProbeABIDetectorTest::testContainsQtCore(unix3)
PASS   : ProbeABIDetectorTest::testContainsQtCore(unix4)
PASS   : ProbeABIDetectorTest::testContainsQtCore(mac1)
PASS   : ProbeABIDetectorTest::testContainsQtCore(mac2)
PASS   : ProbeABIDetectorTest::testContainsQtCore(mac3)
PASS   : ProbeABIDetectorTest::testContainsQtCore(mac4)
PASS   : ProbeABIDetectorTest::testContainsQtCore(mac5)
PASS   : ProbeABIDetectorTest::testContainsQtCore(mac debug1)
PASS   : ProbeABIDetectorTest::testContainsQtCore(mac debug2)
PASS   : ProbeABIDetectorTest::testContainsQtCore(mac qt55 framework)
PASS   : ProbeABIDetectorTest::testContainsQtCore(win1)
PASS   : ProbeABIDetectorTest::testContainsQtCore(win2)
PASS   : ProbeABIDetectorTest::testContainsQtCore(win3)
PASS   : ProbeABIDetectorTest::testContainsQtCore(win debug1)
PASS   : ProbeABIDetectorTest::testContainsQtCore(win debug2)
PASS   : ProbeABIDetectorTest::testContainsQtCore(complex path)
PASS   : ProbeABIDetectorTest::testContainsQtCore(addon1)
PASS   : ProbeABIDetectorTest::testContainsQtCore(addon2)
PASS   : ProbeABIDetectorTest::testContainsQtCore(QT)
PASS   : ProbeABIDetectorTest::testContainsQtCore(prefix)
PASS   : ProbeABIDetectorTest::testContainsQtCore(libQt)
PASS   : ProbeABIDetectorTest::cleanupTestCase()
Totals: 28 passed, 2 failed, 0 skipped, 0 blacklisted
********* Finished testing of ProbeABIDetectorTest *********

      Start  8: remotemodeltest
 8/21 Test  #8: remotemodeltest ..................   Passed    0.13 sec
      Start  9: signalspycallbacktest
 9/21 Test  #9: signalspycallbacktest ............   Passed    0.03 sec
      Start 10: multithreadingtest
10/21 Test #10: multithreadingtest ...............   Passed   13.85 sec
      Start 11: translatortest
11/21 Test #11: translatortest ...................   Passed    0.08 sec
      Start 12: widgettest
12/21 Test #12: widgettest .......................   Passed    0.12 sec
      Start 13: actiontest
13/21 Test #13: actiontest .......................   Passed    0.11 sec
      Start 14: metaobjecttest
14/21 Test #14: metaobjecttest ...................   Passed    0.01 sec
      Start 15: propertysyncertest
15/21 Test #15: propertysyncertest ...............   Passed    0.03 sec
      Start 16: propertybindertest
16/21 Test #16: propertybindertest ...............   Passed    0.03 sec
      Start 17: propertyadaptortest
17/21 Test #17: propertyadaptortest ..............   Passed    0.01 sec
      Start 18: propertymodeltest
18/21 Test #18: propertymodeltest ................   Passed    0.01 sec
      Start 19: fontdatabasemodeltest
19/21 Test #19: fontdatabasemodeltest ............   Passed    0.02 sec
      Start 20: qmlsupporttest
20/21 Test #20: qmlsupporttest ...................   Passed    0.06 sec
      Start 21: quickinspectortest
21/21 Test #21: quickinspectortest ...............   Passed    2.48 sec

76% tests passed, 5 tests failed out of 21

Total Test time (real) =  17.57 sec

The following tests FAILED:
	  1 - connectiontest-preload (Failed)
	  2 - connectiontest-style (Failed)
	  3 - connectiontest-gdb (Failed)
	  4 - attachtest-gdb (OTHER_FAULT)
	  7 - probeabidetectortest (Failed)
Errors while running CTest
Makefile:86: recipe for target 'test' failed
make[2]: *** [test] Error 8
make[2]: Leaving directory '/build/gammaray-2.4.1/obj-qt5'
dh_auto_test: make -j1 test ARGS+=-j1 returned exit code 2
debian/rules:29: recipe for target 'override_dh_auto_test' failed
make[1]: *** [override_dh_auto_test] Error 2
```

**Comment by xhaakon** [comment]
The failure in `QuickInspectorTest::testFetchingPreview()` is caused by a Qt bug [fixed](https://codereview.qt-project.org/146298) in 5.7.

**Comment by krf** [comment]
@xhaakon: Thanks!

@LocutusOfBorg: 

```
FAIL!  : ProbeABIDetectorTest::testDetectExecutable() Compared values are not the same
   Actual   (abi.id())                          : "qt5.5-i686"
   Expected (QStringLiteral(GAMMARAY_PROBE_ABI)): "qt5.5-x86_64"
   Loc: [/build/gammaray-2.4.1/tests/probeabidetectortest.cpp(48)]
FAIL!  : ProbeABIDetectorTest::testDetectProcess() Compared values are not the same
   Actual   (abi.id())                          : "qt5.5-i686"
   Expected (QStringLiteral(GAMMARAY_PROBE_ABI)): "qt5.5-x86_64"
   Loc: [/build/gammaray-2.4.1/tests/probeabidetectortest.cpp(56)]
```

... looks special. You say you're building on i386, but above output suggests `CMAKE_SYSTEM_PROCESSOR` returned x86_64 for you.

Are you cross-compiling for 64 bit on a 32 bit host?

**Comment by LocutusOfBorg** [comment]
Hi @krf 
Are you cross-compiling for 64 bit on a 32 bit host?

quite the opposite I would say.
Debian machines building i386 packages are actually amd64 host machines

```
Version: 2.4.0-1
Source Version: 2.4.0-1
Distribution: sid
Machine Architecture: amd64
Host Architecture: i386
Build Architecture: i386
```
same for Ubuntu
https://launchpad.net/ubuntu/+source/gammaray/2.4.0-1build1/+build/9894591

so, I would say you are right
https://buildd.debian.org/status/fetch.php?pkg=gammaray&arch=i386&ver=2.4.0-1&stamp=1465895008

**Comment by krf** [comment]
Thanks for the confirmation. This gives us some hints.

**Comment by winterz** [comment]
you need to build with a cmake toolchain if you are cross-compiling.
I see you are running cmake as:
cmake .. -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_INSTALL_RPATH_USE_LINK_PATH=FALSE

So you need to also pass something like -DCMAKE_TOOLCHAIN_FILE=cmake/Toolchain-Linux-GCC-i686.cmake

You'll find more information on this in the wiki:
http://github.com/KDAB/GammaRay/wiki/Cross-compiling-GammaRay

**Comment by LocutusOfBorg** [comment]
thanks! I'll leave the fix to @xhaakon since there is also the new qt that needs to go in unstable

**Comment by pasnox** [comment]
Initial issue fixed with 3d0f04cf52d87e461a994a3fb0ead178a00520d6
